### PR TITLE
[ios] #5039 refactored MGLUserLocationAnnotationView

### LINKED ISSUE #5039: MGLUserLocationAnnotationView should inherit from MGLAnnotationView
Now that MGLAnnotationView has landed in #4801, we should make MGLUserLocationAnnotationView, inherit from it. This will be important for consistency once we implement #5037.

/cc @boundsj @friedbunny

### PR DESCRIPTION
**Pull request by frederoni**
Fixes #5039 and #5037
`MGLUserLocationAnnotationView` now inherits from `MGLAnnotationView`.
`MGLMapView` calls `-[MGLMapViewDelegate mapView:viewForAnnotation:]` for `MGLUserLocation` to support a custom user dot. If the implementer still wants the default user dot, they should return nil in case annotation equals `mapView.userLocation`

### CODE REVIEW

**Comment by frederoni** [comment]
**This proposal is very much work in progress.**

The UI part of the code in `MGLUserLocationAnnotationView` has been moved to `MGLFaux3DUserLocationAnnotationView` (naming suggestions are more than welcome) leaving only crucial parts like accessibility and references to userLocation and mapView in the base class. With this abstraction we can untangle some calls from `MGLMapView` like updating the visibility of the haloLayer and setting up layers. This can now be achieved by overriding `-didUpdateUserLocation:` from a `MGLUserLocationAnnotationView`subclass. It is a bit redundant since map view already informs the delegate with `-mapView:didUpdateUserLocation:` but it makes it a bit cleaner when you subclass a user annotation view.

`MGLUserLocationAnnotationView` would become public and the new `MGLFaux3DUserLocationAnnotationView` or whatever we decide to name it would be private.

👀 @friedbunny @boundsj @1ec5

File: platform/ios/app/MBXUserLocationAnnotationView.h [hunk 1]
```diff
@@ -0,0 +1,5 @@
+#import "MGLUserLocationAnnotationView.h"
```
Comment (1ec5) [inline platform/ios/app/MBXUserLocationAnnotationView.h]: You should be able to just import the Mapbox umbrella header or module. But this PR is missing changes to Mapbox.h and the project file that would expose MGLUserLocationAnnotationView.h publicly.

**Comment by 1ec5** [comment]
MGLUserLocationAnnotationView.h needs to be exposed publicly. See [these instructions](https://github.com/mapbox/mapbox-gl-native/blob/master/platform/ios/DEVELOPING.md#adding-a-source-code-file) for details.

File: platform/ios/src/MGLUserLocationAnnotationView.h [hunk 2]
```diff
@@ -9,15 +10,15 @@ NS_ASSUME_NONNULL_BEGIN
 @class MGLUserLocation;
 
 /** View representing an MGLUserLocation on screen. */
-@interface MGLUserLocationAnnotationView : UIView
+@interface MGLUserLocationAnnotationView : MGLAnnotationView
 
 @property (nonatomic, weak) MGLMapView *mapView;
-@property (nonatomic) MGLUserLocation *annotation;
-@property (nonatomic, readonly, nullable) CALayer *haloLayer;
+@property (nonatomic, readonly, nullable) MGLUserLocation *userLocation;
```
Comment (1ec5) [inline platform/ios/src/MGLUserLocationAnnotationView.h]: Now that this header is public, it needs documentation for each symbol.

File: platform/ios/src/MGLUserLocationAnnotationView.h [hunk 3]
```diff
@@ -9,15 +10,15 @@ NS_ASSUME_NONNULL_BEGIN
...
 
-- (instancetype)initInMapView:(MGLMapView *)mapView NS_DESIGNATED_INITIALIZER;
-- (void)setupLayers;
+- (instancetype)initInMapView:(MGLMapView *)mapView userLocation:(MGLUserLocation *)userLocation;
```
Comment (1ec5) [inline platform/ios/src/MGLUserLocationAnnotationView.h]: I know we’ve done it this way for a long time, but I’m not sure we want to make public an API that’s based on a backpointer. It’s just too tempting for subclasses to abuse this backpointer in ways that’ll leak memory. Instead, we should probably figure out a way to push the necessary state to this class.

File: platform/ios/src/MGLFaux3DUserLocationAnnotationView.m [hunk 4]
```diff
@@ -0,0 +1,522 @@
+#import "MGLFaux3DUserLocationAnnotationView.h"
+
+#import "MGLMapView.h"
+#import "MGLUserLocation_Private.h"
```
Comment (1ec5) [inline platform/ios/src/MGLFaux3DUserLocationAnnotationView.m]: Does this file use MGLUserLocation_Private.h for anything, or can it import MGLUserLocation instead? Ideally, a developer could theoretically implement anything that’s in this class in their own custom subclass of MGLUserLocationAnnotationView.

**Comment by 1ec5** [comment]
Since you’re exposing MGLUserLocationAnnotationView publicly for the first time, you’ll also need to add it to jazzy’s table of contents. In d0766120ca2513e44584f12014283de4c361372b, I updated the contributing guide to point this out. Feel free to add anything that appears to be missing.

File: platform/ios/app/MBXUserLocationAnnotationView.m [hunk 5]
```diff
@@ -9,46 +10,168 @@ @interface MBXUserLocationAnnotationView()
 
 @implementation MBXUserLocationAnnotationView
 
+- (instancetype)initWithFrame:(CGRect)frame
+{
+    self = [super initWithFrame:frame];
```
Comment (1ec5) [inline platform/ios/app/MBXUserLocationAnnotationView.m]: Nit: The structure of this implementation works, but we should adhere to the format used in the initializer code snippet. Consistency with initializers makes it easier for us to spot errors like missing assignments to self or return statements.

```objc
if (self = [super initWithFrame:frame]) {
    self.backgroundColor = [UIColor clearColor];
}
return self;
```

**Comment by 1ec5** [comment]
What’s a good way to address https://github.com/mapbox/mapbox-gl-native/issues/5571#issuecomment-234372579? Do we need to expose a new property on MGLUserLocationAnnotationView for the tappable layer or Bézier path?

**Comment by frederoni** [comment]
The frame is correct and the halo layer is drawn out of bounds.
screenshot 2016-07-25 09 49 46
Keep it simple and just hit-test the view's frame?

**Comment by 1ec5** [comment]
The problem is that the view is animated using implicit animation. As the user moves in a straight line, the frame jumps to the CLLocation while the visible dot animates smoothly. The user should be able to tap the dot without having to guess where it’ll wind up.

**Comment by friedbunny** [comment]
I rebased this on master and then force-pushed. This PR targets the v3.3.0 release branch, so I’m closing it in favor of https://github.com/mapbox/mapbox-gl-native/pull/5882.

**Pull request closed without merging**